Tests: standardize the way we check for thrown exceptions

sijie123 opened the pull request:
We have many different ways of checking for thrown exceptions. These are not standardized, and make it difficult for a reader because they have to learn the different ways of throwing exceptions.

Let's standardize the way we test error handling behavior, specifically towards using Assert#assertThrows.

pyokagan commented:
Range diff with #959 v8:

```
 1:  92df84c5 <  -:  -------- VersionTest, VersionedAddressBookTest: Use 8-spaces indent for parameter lists
 2:  59794080 <  -:  -------- TestApp: Move SAVE_LOCATION_FOR_TESTING into AddressBookSystemTest
 3:  34d0c12f <  -:  -------- TestApp: Move DEFAULT_PREF_LOCATION_FOR_TESTING into AddressBookSystemTest
 4:  99272ead <  -:  -------- TestApp: Remove unused default constructor
 5:  762c3567 <  -:  -------- TestApp: Change visibility of initialDataSupplier to private
 6:  39f8435d <  -:  -------- TestApp: Remove null supplier for initialDataSupplier
 7:  4adcd0d7 <  -:  -------- CheckStyle: Allow exceptions for JUnit 5 test annotations
 8:  1d975cac <  -:  -------- Assert: Use JUnit 5 Assert#assertThrows
 9:  1e5e3136 <  -:  -------- Assert: Remove unnecessary VoidCallable interface
10:  7e9b8dd0 =  1:  b65281b5 Assert: Standardize assert statements to use static import
11:  35306231 =  2:  6984f217 LogicManagerTest: Split assertCommandBehavior into assertCommandSuccess and assertCommandFailure
12:  752921d0 !  3:  4bbc7ace Thrown exception checking: Migrate to use JUnit 5 AssertThrows
    @@ -1,6 +1,6 @@
     Author: Si Jie <sijie123@gmail.com>
     
    -    Thrown exception checking: Migrate to use JUnit 5 AssertThrows
    +    Thrown exception checking: Migrate to use AssertThrows
     
         We use a mix of 4 different ways to test exception throwing. We use:
         * try/catch - catch and ensure we receive the correct exception.
    @@ -14,10 +14,10 @@
         one style of implementation.
     
         The try/catch method is implemented differently in each file that
    -    requires it and violates the DRY principle. In JUnit 5, rules are
    -    unsupported, removing compatability with the ExpectedException rule.
    -    This means that we should strive to use only the assertThrows API. In
    -    Assert.java, we implement custom methods that wraps over
    +    requires it and violates the DRY principle. JUnit#assertThrows
    +    provides a shorter syntax and is easier to read as compared to
    +    ExpectedException. We should thus aim to use the assertThrows style.
    +    In Assert.java, we implement custom methods that wraps over
         JUnit#assertThrows to provide additional functionality such as testing
         for correctness of error messages shown to users. With
         Assert#assertThrows providing a superset of required functionality, it
    @@ -57,9 +57,7 @@
          public void versionParsing_wrongVersionString_throwIllegalArgumentException() {
     -        thrown.expect(IllegalArgumentException.class);
     -        Version.fromString("This is not a version string");
    -+        assertThrows(IllegalArgumentException.class, () ->
    -+                Version.fromString("This is not a version string")
    -+        );
    ++        assertThrows(IllegalArgumentException.class, () -> Version.fromString("This is not a version string"));
          }
      
          @Test
    @@ -350,14 +348,13 @@
     -            assertEquals(expectedSelectedPerson, actualModel.getSelectedPerson());
     -            assertEquals(expectedCommandHistory, actualCommandHistory);
     -        }
    --    }
     +        assertThrows(CommandException.class, expectedMessage, () -> command.execute(actualModel, actualCommandHistory));
     +        assertEquals(expectedAddressBook, actualModel.getAddressBook());
     +        assertEquals(expectedFilteredList, actualModel.getFilteredPersonList());
     +        assertEquals(expectedSelectedPerson, actualModel.getSelectedPerson());
     +        assertEquals(expectedCommandHistory, actualCommandHistory);
    - 
    -+    }
    +     }
    +-
          /**
           * Updates {@code model}'s filtered list to show only the person at the given {@code targetIndex} in the
           * {@code model}'s address book.
13:  6142a427 !  4:  af423408 StringUtilTest: Simplify assertions
    @@ -2,15 +2,14 @@
     
         StringUtilTest: Simplify assertions
     
    -    StringUtilTest uses JUnit 4, which has poor support for asserting
    -    thrown exceptions. Hence, we create our own method
    +    In StringUtilTest, we create our own method
         StringUtilTest#assertExceptionThrown to abstract the idea of asserting
         thrown exceptions.
     
    -    Tests can directly use Assert#assertThrows to check for the correct
    -    exception.
    +    As per a previous commit, tests should directly use
    +    Assert#assertThrows to test for the correct exception.
     
    -    Let's migrate all tests to use Assert#assertThrows.
    +    Let's migrate all tests to use Assert#assertThrows instead.
     
         As a result, we no longer need StringUtilTest#assertExceptionThrown.
     
    @@ -58,8 +57,7 @@
     -        thrown.expect(exceptionClass);
     -        errorMessage.ifPresent(message -> thrown.expectMessage(message));
     -        StringUtil.containsWordIgnoreCase(sentence, word);
    -+        assertThrows(NullPointerException.class, ()
    -+            -> StringUtil.containsWordIgnoreCase("typical sentence", null));
    ++        assertThrows(NullPointerException.class, () -> StringUtil.containsWordIgnoreCase("typical sentence", null));
          }
      
          @Test
14:  5725ab1e !  5:  448aa7cf IndexTest: Simplify assertions
    @@ -2,19 +2,25 @@
     
         IndexTest: Simplify assertions
     
    -    IndexTest uses JUnit 4, which has poor support for asserting thrown
    -    exceptions. Hence, we create our own method
    -    IndexTest#assertCreateFailure to encapsulate the idea of asserting
    -    thrown exceptions.
    +    In IndexTest, we create our own method IndexTest#assertCreateFailure
    +    to encapsulate the idea of asserting thrown exceptions. Tests call
    +    the respective assertCreateZeroBasedFailure or
    +    assertCreateOneBasedFailure that acts as a wrapper to
    +    assertCreateFailure.
     
    -    Tests can directly use Assert#assertThrows to check for the correct
    -    exception. Hence, there is no longer a need to encapsulate this idea
    -    with IndexTest#assertCreateFailure.
    +    As per a previous commit, tests should directly directly use
    +    Assert#assertThrows to test for the correct exception.
    +
    +    Let's migrate all tests to use Assert#assertThrows instead.
    +
    +    As a result, we no longer need IndexTest#assertCreateFailure. Tests
    +    can also directly call Assert#assertThrows, so abstraction using
    +    assertCreateZeroBasedFailure and assertCreateOneBasedFailure methods
    +    provide little benefit.
     
         Let's:
         * remove the IndexTest#assertCreateFailure method
    -    * replace calls to IndexTest#assertCreateFailure with
    -    Assert#assertThrows(Throwable, Executable) instead
    +    * remove the IndexTest#assertCreateZero/OneBasedFailure methods
     
         This will simplify our code base and adhere closer to the DRY
         principle instead of re-implementing methods.
    @@ -28,18 +34,14 @@
      import static org.junit.Assert.assertTrue;
     +import static seedu.address.testutil.Assert.assertThrows;
      
    --import org.junit.Test;
    -+import org.junit.jupiter.api.Test;
    - 
    - public class IndexTest {
    + import org.junit.Test;
      
    +@@
          @Test
          public void createOneBasedIndex() {
              // invalid index
     -        assertCreateOneBasedFailure(0);
    -+        assertThrows(IndexOutOfBoundsException.class, () ->
    -+                Index.fromOneBased(0)
    -+        );
    ++        assertThrows(IndexOutOfBoundsException.class, () -> Index.fromOneBased(0));
      
              // check equality using the same base
              assertEquals(1, Index.fromOneBased(1).getOneBased());
    @@ -48,9 +50,7 @@
          public void createZeroBasedIndex() {
              // invalid index
     -        assertCreateZeroBasedFailure(-1);
    -+        assertThrows(IndexOutOfBoundsException.class, () ->
    -+                Index.fromZeroBased(-1)
    -+        );
    ++        assertThrows(IndexOutOfBoundsException.class, () -> Index.fromZeroBased(-1));
      
              // check equality using the same base
              assertEquals(0, Index.fromZeroBased(0).getZeroBased());
15:  097c5305 <  -:  -------- JUnit Assert: Migrate to JUnit 5 API
16:  4cbe6140 <  -:  -------- JUnit Assume: Migrate to JUnit 5 Assume API
17:  885e7b75 <  -:  -------- JUnit: Upgrade to JUnit 5.4.0
18:  0993b780 <  -:  -------- JUnit: Migrate to use JUnit 5 test runner
19:  188c292d <  -:  -------- TemporaryFolder: Create helper class to manage temporary directory contents
20:  089b9903 <  -:  -------- TemporaryFolder: Migrate temporary file creation to DirectoryInitUtil#createTemporaryFileInFolder
21:  bbbcc4ca <  -:  -------- TemporaryFolder: Copy test data files to temporary folder before running tests
22:  84b7efc5 <  -:  -------- TemporaryFolder: Remove redundant addToTestDataPathIfNotNull
23:  8d02a914 <  -:  -------- TemporaryFolder: Migrate getFilePathInSandBoxFolder to Path#resolve
24:  9d0aae5c <  -:  -------- TestUtil: Remove getFilePathInSandboxFolder
25:  ce87de2b <  -:  -------- Gitignore: Remove sandbox folder
26:  2c7300d5 <  -:  -------- JsonUtilTest: Rename SERIALIZATION_FILE to serializationFile
27:  6541db0c <  -:  -------- UiTest: Create StageExtension and UiPartExtension
28:  b15661fb <  -:  -------- UiTest: Migrate StageRule to StageExtension and UiPartRule to UiPartExtension
29:  03ec974c <  -:  -------- UiPartTest: Migrate to JUnit 5 Temporary Directory
30:  180d7519 <  -:  -------- ClockRule: Migrate to ClockExtension
31:  f091de8f <  -:  -------- AB4_Test: Remove unnecessary exception declaration
32:  3de17e1e <  -:  -------- Checkstyle: Remove exceptions for JUnit 4 annotations
33:  01cda734 <  -:  -------- Build.gradle: Remove JUnit 4 libraries
```

pyokagan commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -93,68 +89,65 @@ public void execute_storageThrowsIoException_throwsCommandException() throws Exc
-    private void assertCommandSuccess(String inputCommand, String expectedMessage, Model expectedModel) {
-        assertCommandBehavior(null, inputCommand, expectedMessage, expectedModel);
+    private void assertCommandSuccess(String inputCommand, String expectedMessage,
+            Model expectedModel) throws Exception {
```
This is just a utility function, should we be listing out the exceptions fully here? i.e. `throws ParseException, CommandException`.

pyokagan commented on src/test/java/seedu/address/storage/JsonSerializableAddressBookTest.java:
```diff
@@ -37,17 +33,15 @@ public void toModelType_typicalPersonsFile_success() throws Exception {
                 JsonSerializableAddressBook.class).get();
-        thrown.expect(IllegalValueException.class);
-        dataFromFile.toModelType();
+        assertThrows(IllegalValueException.class, () -> dataFromFile.toModelType());
```
`dataFromFile::toModelType`?

sijie123 commented:
Oops. Accidentally CIHR-ed with a bug in it. Please ignore v3 and look at v4 instead.
Thanks @pyokagan for the review. I've updated the PR with 2 things:
1. Yup, I agree dataFromFile::toModelType is semantically clearer, and listing out the 2 exceptions in the helper method makes sense. I've updated them. As for "propagating the exception", I see where you're coming from. However I don't have a nice way of saying "ensure that tests fail if the tested methods throw an exception". I think the existing line is probably the easiest to understand...
2. I've squashed in [v8 31/33] from #959. They should really have been part of this PR.

pyokagan commented:
@sijie123 Thanks, I think you'll need to squash this into [3/3] as well though:

```diff
diff --git a/src/test/java/seedu/address/logic/parser/ParserUtilTest.java b/src/test/java/seedu/address/logic/parser/ParserUtilTest.java
index 5291aac9..9dfdb20a 100644
--- a/src/test/java/seedu/address/logic/parser/ParserUtilTest.java
+++ b/src/test/java/seedu/address/logic/parser/ParserUtilTest.java
@@ -150,12 +150,12 @@ public class ParserUtilTest {
     }
 
     @Test
-    public void parseTag_null_throwsNullPointerException() throws Exception {
+    public void parseTag_null_throwsNullPointerException() {
         assertThrows(NullPointerException.class, () -> ParserUtil.parseTag(null));
     }
 
     @Test
-    public void parseTag_invalidValue_throwsParseException() throws Exception {
+    public void parseTag_invalidValue_throwsParseException() {
         assertThrows(ParseException.class, () -> ParserUtil.parseTag(INVALID_TAG));
     }
 
@@ -173,12 +173,12 @@ public class ParserUtilTest {
     }
 
     @Test
-    public void parseTags_null_throwsNullPointerException() throws Exception {
+    public void parseTags_null_throwsNullPointerException() {
         assertThrows(NullPointerException.class, () -> ParserUtil.parseTags(null));
     }
 
     @Test
-    public void parseTags_collectionWithInvalidTags_throwsParseException() throws Exception {
+    public void parseTags_collectionWithInvalidTags_throwsParseException() {
         assertThrows(ParseException.class, () -> ParserUtil.parseTags(Arrays.asList(VALID_TAG_1, INVALID_TAG)));
     }
 
```

sijie123 commented:
Hmm... How did I miss that one out ><
Updated.

pyokagan commented:
CIHR pls.

Zhiyuan-Amos commented:
[1/3]

JsonAdaptedTest#95-100

```java
    @Test
    public void toModelType_nullAddress_throwsIllegalValueException() {
        JsonAdaptedPerson person = new JsonAdaptedPerson(VALID_NAME, VALID_PHONE, VALID_EMAIL, null, VALID_TAGS);
        String expectedMessage = String.format(MISSING_FIELD_MESSAGE_FORMAT, Address.class.getSimpleName());
        Assert.assertThrows(IllegalValueException.class, expectedMessage, person::toModelType);
    }
```

This test (along with a few others) are still using `Assert.assertThrows(...)` instead of using the static import, is this intended?

sijie123 commented:
> This test (along with a few others) are still using `Assert.assertThrows(...)` instead of using the static import, is this intended?

Thanks for the catch. I've gone one more round to make sure that all these static imports are included. Should be good now.

> [3/3]
> 
> > In Assert.java, we implement custom methods that wraps over
> > JUnit#assertThrows to provide additional functionality such as testing
> > for correctness of error messages shown to users.
> 
> I don't think so :P https://github.com/se-edu/addressbook-level4/blob/master/src/test/java/seedu/address/testutil/Assert.java

Thanks. I cherry-picked this one over, so I forgot that I had not updated Assert.java in this PR.

sijie123 commented:
I meant to say that we use JUnit's Assertions.assertThrows [in some places](https://github.com/se-edu/addressbook-level4/pull/1007/commits/d85533a2ff0fd4ab567b1705abc5780837f4e8d5#diff-f63eb345295fd55d62dbfd9272ecc675), and that we should change these instances to Assert#assertThrows.

That said, 
> The try/catch method is implemented differently in each file that
requires it and violates the DRY principle. JUnit#assertThrows
provides a shorter syntax and is easier to read as compared to
ExpectedException. We should thus aim to use the assertThrows style.

I do think that this has an unnecessary reference to JUnit.
"The try/catch method is implemented differently in each file that
requires it and violates the DRY principle. **The assertThrows style**
provides a shorter syntax and is easier to read as compared to
ExpectedExceptions or try/catch. We should thus aim to use the assertThrows style."

pyokagan commented:
@sijie123 
> I meant to say that we use JUnit's Assertions.assertThrows in some places, and that we should change these instances to Assert#assertThrows.

I thought @Zhiyuan-Amos pointed out in https://github.com/se-edu/addressbook-level4/pull/1007#pullrequestreview-230635741 that JUnit 4 doesn't have assertThrows?

sijie123 commented:
> @sijie123
> 
> > I meant to say that we use JUnit's Assertions.assertThrows in some places, and that we should change these instances to Assert#assertThrows.
> 
> I thought @Zhiyuan-Amos pointed out in [#1007 (review)](https://github.com/se-edu/addressbook-level4/pull/1007#pullrequestreview-230635741) that JUnit 4 doesn't have assertThrows?

Yup, JUnit 4 doesn't have assertThrows. I'll need to be careful not to imply that ><

This Assertions.assertThrows exists because our existing code already uses it in some places (this [commit](https://github.com/se-edu/addressbook-level4/commit/ece049f7a66b3f3b4a5a8a4e65b173e84464019d) introduced one).

pyokagan commented:
> I thought @Zhiyuan-Amos pointed out in [#1007 (review)](https://github.com/se-edu/addressbook-level4/pull/1007#pullrequestreview-230635741) that JUnit 4 doesn't have assertThrows?

Oh I see, we are using JUnit5's assertThrows in one place. Well, the wonders of having two versions of JUnit at a single time I guess.
